File: competitive-programming/src/edu/algos/li15_tree/_Bst_L113_PathSumII.java
package edu.algos.li15_tree;

import java.util.ArrayList;
import java.util.List;

import edu.algos.li30_model.TreeNode;


public class _Bst_L113_PathSumII {

	public static void main(String[] args) {
		_Bst_L113_PathSumII obj = new _Bst_L113_PathSumII();
		
		TreeNode root = new TreeNode(1);
		root.left = new TreeNode(3);
		root.right = new TreeNode(4);
		root.left.left = new TreeNode(2);
        root.left.right = new TreeNode(6);
        root.left.right.left = new TreeNode(5);
		
		System.out.println(obj.pathSum(root, 6));
	}

	/*
	 * Problem: https://leetcode.com/problems/path-sum-ii/
	 * Companies: Amazon
     * Idea: Backtracking algo. Add the node to list when node processing started, remove while exiting.
	 * */
    // time: O(n)
    // space: O(n)
	public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<List<Integer>> result = new ArrayList<>();
        pathSum(root, targetSum, new ArrayList<Integer>(), result);
        return result;
    }
    
    private void pathSum(TreeNode root, int targetSum, List<Integer> list, List<List<Integer>> result) {
        if(root == null)  
            return;
        
        list.add(root.val);
        if(isLeaf(root) && targetSum == root.val)
            result.add(new ArrayList<Integer>(list));
        
        pathSum(root.left, targetSum - root.val, list, result);
        pathSum(root.right, targetSum - root.val, list, result);
        list.remove(list.size()-1);
    }

    private boolean isLeaf(TreeNode root) {
        return root.left == null && root.right == null;
    }
}

File: dsa/patterns/10_trees/bst/python/bst_l113__path_sum_ii.py
# Tree / Backtracking
# Problem: https://leetcode.com/problems/path-sum-ii/
# Companies: Amazon
# Idea:
# 1. Use backtracking to explore all root-to-leaf paths.
# 2. Add node value to path when entering, remove when exiting.
# 3. If leaf node and sum equals target, add a copy of path to result.
# Time: O(n)
# Space: O(n) recursion stack + path storage

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def path_sum(root, target_sum):
    result = []

    def dfs(node, remaining_sum, path):
        if not node:
            return

        # Add current node to path
        path.append(node.val)

        # Check if leaf and sum matches
        if not node.left and not node.right and remaining_sum == node.val:
            result.append(list(path))  # append a copy

        # Recurse left and right
        dfs(node.left, remaining_sum - node.val, path)
        dfs(node.right, remaining_sum - node.val, path)

        # Backtrack
        path.pop()

    dfs(root, target_sum, [])
    return result


# -------- Test --------
root = TreeNode(1)
root.left = TreeNode(3)
root.right = TreeNode(4)
root.left.left = TreeNode(2)
root.left.right = TreeNode(6)
root.left.right.left = TreeNode(5)

print(path_sum(root, 6))  # Expected: [[1,3,2], [1,3,6,5]] depending on tree
